test/csuite/wt7989_compact_checkpoint/main.c

```diff
@@ -0,0 +1,286 @@
+    if (!opts->preserve)
+        testutil_clean_work_dir(home_cv);
+
+    testutil_cleanup(opts);
+
+    return (EXIT_SUCCESS);
+}
+
+static void
+run_test(const char *home, const char *uri, bool stress_test)
+{
+    struct thread_data td;
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    pthread_t thread_checkpoint, thread_compact;
+    uint64_t file_sz_after, file_sz_before;
+
+    testutil_make_work_dir(home);
+    testutil_check(wiredtiger_open(home, NULL, conn_config, &conn));
+
+    if (stress_test) {
+        /*
+         * Set WT_TIMING_STRESS_CHECKPOINT_SLOW flag. It adds 10 seconds sleep before each
+         * checkpoint.
+         */
+        set_timing_stress_checkpoint(conn);
+    }
+
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    /* Create and populate dataset. Checkpoint the data after that. */
+    testutil_check(session->create(session, uri, table_config));
+
+    populate(session, uri);
+    testutil_check(session->checkpoint(session, NULL));
+
+    /*
+     * Remove 1/3 of data from the middle of the key range. To let compact relocate blocks from the
+     * end of the file.
+     */
+    remove_records(session, uri);
+
+    file_sz_before = get_file_size(session, uri);
+
+    /* Create and initialize conditional variable. */
+    td.conn = conn;
+    td.uri = uri;
+    if (stress_test) {
+        td.cond = NULL;
+        td.session_cond = NULL;
+    } else {
+        td.session_cond = (WT_SESSION_IMPL *)session;
+        testutil_check(__wt_cond_alloc(td.session_cond, "compact operation", &td.cond));
+    }
+
+    /* Spawn checkpoint and compact threads. */
+    testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+    testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+    (void)pthread_join(thread_checkpoint, NULL);
+    (void)pthread_join(thread_compact, NULL);
+
+    file_sz_after = get_file_size(session, uri);
+
+    /* Cleanup */
+    if (!stress_test) {
+        __wt_cond_destroy(td.session_cond, &td.cond);
+        td.session_cond = NULL;
+        td.cond = NULL;
+    }
+
+    testutil_check(session->close(session, NULL));
+
+    /* Check if there's at least 10% compaction. */
+    testutil_assertfmt(file_sz_before * 0.9 > file_sz_after,
+      "Compact failed to reduce file size by at least 10%%\n"
+      "Compressed file size MB: %f\n"
+      "Original file size MB: %f\n",
+      file_sz_after / (1024.0 * 1024), file_sz_before / (1024.0 * 1024));
+}
+
+static void *
+thread_func_compact(void *arg)
+{
+    struct thread_data *td;
+    WT_SESSION *session;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. */
+        __wt_sleep(1, 0);
+
+        /* Wake up the checkpoint thread. */
+        __wt_cond_signal(td->session_cond, td->cond);
```

Comment:
I think the use of `td->session_cond` is incorrect. The API doesn't ask for the session pointer for the thread waiting for this condition variable. I believe we just need to pass the `session` from the current thread.

Reply:
Done